### PR TITLE
add torch.diff support for n greater than 1

### LINKED ISSUE #54853: `diff` higher order
Requesting feature.

```python
RuntimeError: diff only supports n = 1 currently. Please file an issue at 
https://github.com/pytorch/pytorch/issues/new?assignees=&labels=&template=feature-request.md 
if your use case requires supporting higher-order differences
```

Current workaround:

```python
def torch_diff(x, n, dim=-1):
    out = torch.diff(x, dim=dim)
    if n > 1:
        for _ in range(n - 1):
            out = torch.diff(out, dim=dim)
    return out
```

cc @mruberry @rgommers @heitorschueroff

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67260

Addressing 54853

Differential Revision: [D31930294](https://our.internmc.facebook.com/intern/diff/D31930294)

### CODE REVIEW

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:771]: nit: `(self.dtype() == at::kBool;)` for readability

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:763]: Comment here

Code formatting nit: `(n == 0) {`

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:777]: Nice bool handling

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:800]: This check is a little severe but that seems OK

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:806]: nit: `const auto`

File: aten/src/ATen/native/ReduceOps.cpp
Comment (mruberry) [inline aten/src/ATen/native/ReduceOps.cpp:814]: Nice pattern of using `_out` variants for the final computation

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2576]: Great extension

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2587]: This is very thorough (which is great!) but also computationally expensive (which is less great). How about including `n` in the sample input for each sample and just including 1-2 samples where `n > dim_size`?

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2588]: nit: sometimes it's faster to partial the function like:

```
make_arg = partial(make_tensor, dict(device=device, dtype=dtype, requires_grad=requires_grad))
```

then these lines would be just 

```
make_arg(size_prepend) if size_prepend else None
```

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2590]: Hiding `n` and `dim` back here is tough to read. Passing them as kwargs would be more readable:

```
s = SampleInput(make_arg(size), kwargs={
  "n": n,
  "dim": dim,
  ...   
})
```

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7931]: Really nice reference handling

**Review by mruberry — APPROVED** [review]
Hey @mikaylagawarecki! This looks great. I left a few comments/suggestions for your review. Mostly to consider slimming the sample inputs so the tests don't take so long. 

I also added some additional tests, so let's wait for those to complete.

Thanks for patiently waiting for the review! Moving was a lot more time consuming than I expected. As soon as the tests pass this should be OK to land; ping me if you have any additional questions.

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31930294).

**Pull request closed without merging**
